ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -139,205 +119,215 @@ public Driver(QueryState queryState, QueryInfo queryInfo, HiveTxnManager txnMana
-    preparForCompile(resetTaskIds);
-
-    Compiler compiler = new Compiler(context, driverContext, driverState);
-    QueryPlan plan = compiler.compile(command, deferClose);
-    driverContext.setPlan(plan);
-
-    compileFinished(deferClose);
+  @Override
+  public HiveConf getConf() {
+    return driverContext.getConf();
   }
 
-  private void compileFinished(boolean deferClose) {
-    if (DriverState.getDriverState().isAborted() && !deferClose) {
-      closeInProcess(true);
-    }
+  @Override
+  public CommandProcessorResponse run() throws CommandProcessorException {
+    return run(null, true);
   }
 
-  private void preparForCompile(boolean resetTaskIds) throws CommandProcessorException {
-    driverTxnHandler.createTxnManager();
-    DriverState.setDriverState(driverState);
-    prepareContext();
-    setQueryId();
+  @Override
+  public CommandProcessorResponse run(String command) throws CommandProcessorException {
+    return run(command, false);
+  }
 
-    if (resetTaskIds) {
-      TaskFactory.resetId();
+  private CommandProcessorResponse run(String command, boolean alreadyCompiled) throws CommandProcessorException {
+    try {
+      runInternal(command, alreadyCompiled);
+      return new CommandProcessorResponse(getSchema(), null);
+    } catch (CommandProcessorException cpe) {
+      processRunException(cpe);
+      throw cpe;
     }
   }
 
-  private void prepareContext() throws CommandProcessorException {
-    if (context != null && context.getExplainAnalyze() != AnalyzeState.RUNNING) {
-      // close the existing ctx etc before compiling a new query, but does not destroy driver
-      closeInProcess(false);
-    }
+  private void runInternal(String command, boolean alreadyCompiled) throws CommandProcessorException {
+    DriverState.setDriverState(driverState);
+    setInitialStateForRun(alreadyCompiled);
 
+    // a flag that helps to set the correct driver state in finally block by tracking if
+    // the method has been returned by an error or not.
+    boolean isFinishedWithError = true;
     try {
-      if (context == null) {
-        context = new Context(driverContext.getConf());
+      HiveDriverRunHookContext hookContext = new HiveDriverRunHookContextImpl(driverContext.getConf(),
+          alreadyCompiled ? context.getCmd() : command);
+      runPreDriverHooks(hookContext);
+
+      if (!alreadyCompiled) {
+        compileInternal(command, true);
+      } else {
+        driverContext.getPlan().setQueryStartTime(driverContext.getQueryDisplay().getQueryStartTime());
       }
-    } catch (IOException e) {
-      throw new CommandProcessorException(e);
-    }
 
-    context.setHiveTxnManager(driverContext.getTxnManager());
-    context.setStatsSource(driverContext.getStatsSource());
-    context.setHDFSCleanup(true);
+      // Reset the PerfLogger so that it doesn't retain any previous values.
+      // Any value from compilation phase can be obtained through the map set in queryDisplay during compilation.
```

Comment:
Shouldn't this be the first thing in the life cycle? Or minimally in this method? Previous code paths might already started to use the perf logger.